### PR TITLE
Encode quasar.Data as XML in MarkLogic

### LINKED ISSUE #1372: MarkLogic XML codec for quasar.Data
`quasar.Data` is currently stored in MarkLogic as JSON via naive literal encoding, this needs to be extended to use the XML Data Model (XDM) which supports more data types natively and also likely to be the format most MarkLogic data is in.

### PR DESCRIPTION
**Pull request by wemrysi**
Switches to using XML to encode `quasar.Data` in MarkLogic instead of JSON. This allows for a more uniform handling of planning for MarkLogic as we are focusing on XML support and so can ignore JSON for the time-being.

Both the filesystem side and the planning side have been updated and, while they aren't running in travis yet, I have confirmed locally that all the MarkLogic filesystem tests pass with the new encoding.

Closes #1372.

### CODE REVIEW

File: marklogic/src/main/scala/quasar/physical/marklogic/fs/data.scala [hunk 1]
```diff
@@ -18,22 +18,213 @@ package quasar.physical.marklogic.fs
...
+    def inner: QName => Data => Validation[ErrorMessages, Elem] =
+      name => encodeXml0(innerStr, innerElem, inner)(name)
+
+    encodeXml0(rootStr, rootElem, inner)(ejsonEjson)(data)
+      .fold(_.raiseError[F, Elem], _.point[F])
+  }
+
+  def decodeXml[F[_]: MonadErrMsgs](elem: Elem): F[Data] = {
+    def decodeXml0: Node => Validation[ErrorMessages, Data] = {
+      case DataNode("array", children) =>
```
Comment (alissapajer) [inline marklogic/src/main/scala/quasar/physical/marklogic/fs/data.scala]: @wemrysi is there a reason you chose `String` as the type we match on in the `DataNode.unapply`? Could you make an algebra of case objects for each of these cases instead? Or maybe use some type that already exists? 

```scala
sealed trait DataType
final case object Array extends DataType
// etc.
```

Then we'd know if we're missing a case in the pattern match, and we could avoid `String`, which is so generic.

**Comment by alissapajer** [comment]
@wemrysi question regarding testing - so it looks like there are solid unit tests using scalacheck for `encodeXML` and `toData`. What other tests are you running with marklogic? I'm basically wondering if you think this is well-tested from the marklogic point of view?

**Comment by wemrysi** [comment]
@alissapajer The filesystem side of things is reasonably well-tested via the standard `quasar.fs.*` spec tests (all of which pass for ML). The planner side of things, not so much. I don't have any of the regression tests passing yet for various reasons, but this PR is a prerequisite in order to do so.

File: marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala [hunk 2]
```diff
@@ -64,31 +69,40 @@ object queryfile {
...
-        mod <- plan(shifted).leftMap(mlerr => mlerr match {
-          case InvalidQName(s) =>
-            FileSystemError.planningFailed(lp, QPlanner.UnsupportedPlan(
-              // TODO: Change to include the QScript context when supported
-              LogicalPlan.ConstantF(Data.Str(s)), Some(mlerr.shows)))
-        })
-        a   <- WriterT.put(lift(f(mod)).into[S])(phase(mod)).liftM[FileSystemErrT]
+        qs      <- convertToQScriptRead[Fix, QPlan, QSR](listContents)(lp)
+        shifted =  transFutu(qs)(ShiftRead[Fix, QSR, MLQScript].shiftRead(idPrism.reverseGet)(_: QSR[Fix[QSR]]))
+                     .transCata(optimize.applyAll)
```
Comment (alissapajer) [inline marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala]: While it's on my mind, we're working towards having the optimizations/normalizations run automatically multiple times until no further changes to the qscript are detected. But until we're there, it is sometimes necessary to run `optimize.applyAll` sequentially two or three times. For example, we do this in `convertToQScriptRead`. 

So this is totally fine for now, but just a heads up that you might need to simply call `applyAll` again here if the qscript you get out after `shiftRead` can be optimized further.

**Comment by alissapajer** [comment]
@wemrysi LGTM. Once the `DataNode` change is in, I'll merge.

File: marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala (hunk 2 above)
Comment (wemrysi) [inline marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala]: Ok, sounds good. Any way I can detect whether or not this is needed? Or is it trial-and-error at this point?

Comment (alissapajer) [inline marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala]: At this point it's just trial and error. If you're getting qscript that is more complicated than you expect, this is a good first thing to try to see if the qscript gets any better.

**Pull request merged**